### ISSUE TITLE
memory (and possibly observations) leak in io.l5.mesh interface

### ISSUE THREAD

**Issue by edio** [linkerd/linkerd#1906]
## Filing a Linkerd issue ##

Issue Type:

- [x] Bug report
- [ ] Feature request

**What happened**:

Over time namerd memory usage increases if `io.l5d.mesh` interface is used

**What you expected to happen**:

Memory usage won't increase

**How to reproduce it (as minimally and precisely as possible)**:
1. run consul in dev mode
```
consul agent -dev
```

2. run namerd
namerd.yaml
```
admin:
  ip: 127.0.0.1
  port: 9001
namers:
- kind: io.l5d.consul
  prefix: /consul
interfaces:
- kind: io.l5d.mesh
  ip: 0.0.0.0
  port: 4321
telemetry:
  - kind: io.l5d.influxdb
storage:
  kind: io.l5d.inMemory
  namespaces:
    default: |
      /service => /#/consul/dc1;
```
```
./namerd-1.3.7-exec namerd.yaml
```
3. run the following code (all dependencies should be easy to figure out, but let me know if you want me to publish a gradle project)
```
import com.twitter.conversions.time._
import com.twitter.finagle.buoyant.H2
import com.twitter.finagle.service.Backoff
import com.twitter.finagle.util.DefaultTimer
import com.twitter.finagle.{Dtab, Name, Path}
import com.twitter.util.{Await, Closable, Duration, Future, Time, TimeoutException, Timer}
import io.buoyant.interpreter.mesh.Client

object Scenario extends Closable {
  implicit val timer = DefaultTimer
  private[this] val name = Name.Path(Path.read("/$/inet/localhost/4321"))
  private[this] val namespace = Path.read("/default")
  private[this] val label = "io.l5d.mesh"
  private[this] val backoffs = Backoff.exponentialJittered(100.millis, 1.second)
  private[this] val client = H2.client.withParams(H2.client.params).newService(name, label)
  private[this] val interpreter = Client(namespace, client, backoffs, DefaultTimer)

  private[this] val opTimeout = 2.seconds
  private[this] val minPeriod = 10.millis

  override def close(deadline: Time): Future[Unit] = client.close(deadline)
    .raiseWithin(opTimeout, new TimeoutException("failed to close within timeout"))

  def execute(): Future[Unit] = Future.sleep(minPeriod)
    .join(interpreter.bind(Dtab.empty, Path.read(s"/service/consul")).values.toFuture()
        .raiseWithin(opTimeout, new TimeoutException("no response withing timeout"))).unit
}

object ReproduceH2Leak extends App {
  implicit val timer: Timer = DefaultTimer

  @volatile var running: Boolean = true
  def spawn(): Future[Unit] = {
    Scenario.execute().ensure {
      if (running) spawn()
    }
  }

  def load(simultaneousSessions: Int, duration: Duration): Future[Unit] =
    Future {
      println("stressing...")
      running = true
    }.before(
      Future.join(
        Future.join(Future.parallel(simultaneousSessions) {
          spawn()
        }),
        Future.sleep(duration).before(Future {
          println("idling...")
          running = false
        })
      ).unit
    )

  def repeatedly(times: Int, op: => Future[Unit]): Future[Unit] = {
    var step = op
    for (_ <- 1 to times) {
      step = step.before(op)
    }
    step
  }

  val simulation = repeatedly(10, load(10, 45.seconds).before(Future.sleep(15.seconds)))

  Await.ready(simulation.before(Scenario.close()))
  println("stress complete")
}
```
4. observe namerd memory usage growing over time and never falls down, even after idling
https://snapshot.raintank.io/dashboard/snapshot/F4V8BX7JB3IQyxNzDxjnA6iuo9D5iqwM?orgId=2

**Anything else we need to know?**:
I believe, another leak may be somewhere in H2 implementation on client side (not necessarily related). When running stress test I see LEAK warnings from netty
```
SEVERE: LEAK: ByteBuf.release() was not called before it's garbage-collected.
```

Also note, how observation in consul namer is never closed, even if you do just one iteration of the stress. It is not always the case though. I believe, something (in mesh interpreter?) may be holding a reference to lookup activity and keep observing it.

**Environment**:
namerd version: 1.3.7

**Comment by adleong** [comment]
Thank you for the detailed report and repro steps!!!  We will investigate.

**Comment by adleong** [comment]
I've done a little bit of digging and it's starting to look like this may be a bug in H2/gRPC where we're missing a mechanism to reset a response stream that we no longer need.  Since the mesh interface's response streams are effectively infinite, this means that we continue to build up more and more streams over time.

I will continue to investigate and keep this thread updated.

**Comment by adleong** [comment]
@edio please try out #2051 and let me know if it resolves this issue for you

**Comment by edio** [comment]
@adleong 
I tested with the same test I used to reproduce the issue.

Your change fixes the leak, memory consumption is stable. Thanks!

![screenshot_20180711_175529](https://user-images.githubusercontent.com/787439/42584504-9c2b086a-8533-11e8-9431-f6271a52dcd9.png)

Small note (if anyone is interested): this change affects both server and client. If there are any applications/libraries that communicate to namerd via buoyant mesh client and, those are to be updated too to fix the leak.
